File: 406-queue-height/js/406.js
#!/usr/bin/env node

const reconstructQueue = people => {
    // Sort from tallest to shortest
    // Tiebreak: Lower "K" numbers first.
    people.sort( (a, b) => {
        const heightSort = b[0] - a[0];
        const kSort = a[1] - b[1];
        return heightSort == 0 ? kSort : heightSort;
    });

    let q = [people.shift()];
    // Scan the array and insert the person at the correct position
    for (let p = 0; p < people.length; p++) {
        let person = people[p];
        q.splice(person[1], 0, person);
    }
    return q;
};

let test = (expectedResult, actualResult) => {
    console.log("Expected result:", expectedResult);
    console.log("Actual result:", actualResult);
    match = JSON.stringify(actualResult) == JSON.stringify(expectedResult);
    console.log("Pass?", match);
};

const input = [[7,0], [4,4], [7,1], [5,0], [6,1], [5,2]];
let actualResult = reconstructQueue(input);
let expectedResult = [[5,0], [7,0], [5,2], [6,1], [4,4], [7,1]];
test(expectedResult, actualResult);

File: 2019-06/406-queue-height/elixir/406.exs
#! /usr/bin/env elixir

defmodule Solution do
  def reconstruct_queue(people) do
    people
    |> Enum.sort_by(fn x -> {Enum.at(x, 0) * -1, Enum.at(x, 1)} end)
    |> Enum.reduce([], fn person, acc ->
      acc
      |> List.insert_at(Enum.at(person, 1), person)
    end)
  end
end

this_input = [[7, 0], [4, 4], [7, 1], [5, 0], [6, 1], [5, 2]]
expected_result = [[5, 0], [7, 0], [5, 2], [6, 1], [4, 4], [7, 1]]
actual_result = this_input |> Solution.reconstruct_queue()

IO.puts("Expected Result")
expected_result |> IO.inspect()
IO.puts("Actual Result")
actual_result |> IO.inspect()
IO.puts("Equal?")
(expected_result == actual_result) |> IO.inspect()
